src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/UseValueTasksCorrectly.cs

```diff
@@ -0,0 +1,542 @@
+                                continue;
+
+                            case OperationKind.ExpressionStatement:
+                                // Warn! This is a statement. The result should have been used.
+                                operationContext.ReportDiagnostic(invocation.CreateDiagnostic(UnconsumedRule));
+                                return;
+
+                            case OperationKind.Argument:
+                                // The "0.09% case" is passing the result of a call directly as an argument to another method.
+                                // This could later result in a problem, as now there's a parameter inside the callee that's
+                                // holding on to the awaitable, and it could await it twice... but the caller is still correct,
+                                // and this analyzer does not perform inter-method analysis.
+                                var arg = (IArgumentOperation)operation.Parent;
+                                var originalType = arg.Parameter.Type.OriginalDefinition;
+                                if (originalType.Equals(valueTaskType) || originalType.Equals(valueTaskOfTType))
+                                {
+                                    // However, it's really only expected when the parameter type is explicitly a ValueTask{<T>};
+                                    // if it's just, say, a TValue, we're likely on a bad path, such as storing the instance into
+                                    // a collection of some kind, e.g. Dictionary<string, ValueTask>.Add(..., vt).
+                                    var originalParameter = arg.Parameter.OriginalDefinition;
+                                    if (originalParameter.Type.Kind != SymbolKind.TypeParameter)
+                                    {
+                                        return;
+                                    }
+                                }
+                                goto default;
+
+                            default:
+                                // Handle atypical / difficult cases that require more analysis.
+                                HandleAtypicalValueTaskUsage(operationContext, debugType, operation, invocation);
+                                return;
+                        }
+                    }
+                }, OperationKind.Invocation);
+            });
+        }
+
+        /// <summary>Handles more complicated analysis to warn on more complicated erroneous usage patterns.</summary>
+        private static void HandleAtypicalValueTaskUsage(OperationAnalysisContext operationContext, INamedTypeSymbol? debugType, IOperation operation, IInvocationOperation invocation)
+        {
+            if (TryGetLocalSymbolAssigned(operation.Parent, out var valueTypeSymbol, out var startingBlock))
+            {
+                // At this point, it's very likely misuse and we could warn. However, there are a few advanced
```

Comment:
Can you file a tracking issue to attempt converting this to a dataflow analysis based on the DFA framework in the repo and assign it to me?

Reply:
Done: https://github.com/dotnet/roslyn-analyzers/issues/3332. Thanks.